### ISSUE TITLE
Is it possible to use multiple inputs with skorch?

### ISSUE THREAD

**Issue by Mrswolf** [skorch-dev/skorch#672]
Hello
skorch is very useful and it saves me the time of writing training code.

Recently, I need to build a model with multiple inputs, e.g. model(X1, X2, X3). I would like to know there is any way to train this kind of model with skorch.

I read the document and it seems that skorch currently only supports (X, y) .

Thank you.

**Comment by BenjaminBossan** [comment]
Hi, thanks for your kind words.

Indeed, skorch requires to have `X` and `y`. However, `X` can be multiple things. The easiest is for your `X` to be a dictionary. Then the **keys of that dictionary** have to correspond to the **argument names of your module**. Below is an example:

```python

X_dict = {'foo': X1, 'bar': X2, 'baz': X3}

class MyModule(nn.Module):
    def forward(self, foo, bar, baz):  # <- same names as in your dict
        ...

net = NeuralNet(MyModule, ...)
net.fit(X_dict, y)
```

For this to work, `X1`, `X2` and `X3` should have the same length, otherwise we would run into difficulties, e.g. when it comes to how to create mini batches out of these. If you have something more complex (e.g. not the same lengths), then the best way is probably to implement your own [dataset](https://skorch.readthedocs.io/en/stable/user/dataset.html#id1) class.

> I read the document and it seems that skorch currently only supports (X, y)

Indeed, I think we should make a better job at documenting it.

**Comment by BenjaminBossan** [comment]
@Mrswolf Did you make some progress?

**Comment by Mrswolf** [comment]
Yes, I have tried your advice and it works. Now I am implementing my dataset class to deal with unequal lengths.

Thanks again.

**Comment by BenjaminBossan** [comment]
Cool. I'll leave this issue open until we update the docs, since this should really be documented better.

If you have any further questions, feel free to post here.

**Comment by asberk** [comment]
@BenjaminBossan could you help confirm which other formats are acceptable? 

1. ✅ As per the above discussion, it is possible to pass a `dict` to a `skorch.NeuralNet` — whose keys all have the same length — for which the keys are treated as arguments of the `forward` method. 

Will `skorch accept a sub-class of `skorch.dataset.Dataset` that returns multiple inputs? If so, what formats are accepted? For example, say I have a dataset for an image captioning task. There, I have an `input_img`, an `input_caption` and a `target_caption`. 

2. ⛔️ Presently, Skorch does not like it if the dataset returns a 3-tuple, `(input_img, input_caption, target_caption)`. This is expected, as per the quotation from the documentation above.
3. ⛔️ Skorch does not like it if a dataset returns a 3-pair `dict`, `{'image': input_img, 'caption': input_caption, 'target': target_caption}`. This remains true regardless of the keys' names (e.g., you can't simply re-name one of them to `'y'`).
4. 🔸 Skorch *does* accept nested tuples, `((input_img, input_caption), target_caption)`, but then `DataLoader` does not perform batching correctly.
5. ✅ Finally (importantly), Skorch accepts a dataset whose samples are a tuple of `dict` and label tensor, e.g., the dataset `__getitem__` method returns `{'image': input_img, 'caption': input_caption}, target_caption`. In this case, one calls `model.fit(dataset, y=None)` and batching with `DataLoader` seems to work just fine. 

I think the final item is important to note explicitly because in the case of tasks like image captioning, there can be multiple captions per image. Mandating a dictionary input whose values all have the same length means copying a lot of extra image data. Allowing for a dataset that returns a `(dict, Tensor)` tuple mitigates this. I was not able to find explicit mention of this format in the documentation (though it's very possible I've overlooked it). 

If there are other formats that might be amenable to the example I've used above, I would appreciate it if you could point me to them in the documentation. Otherwise, do you think something like this would fit into the [Advanced usage](https://nbviewer.jupyter.org/github/skorch-dev/skorch/blob/master/notebooks/Advanced_Usage.ipynb) notebook?

**Comment by BenjaminBossan** [comment]
>     1. As per the above discussion, it is possible to pass a `dict` to a `skorch.NeuralNet` — whose keys all have the same length — for which the keys are treated as arguments of the `forward` method.

This is correct, just a small correction, you mean "values", not "keys" having the same length. Otherwise, you described the situation pretty well. If you see possible improvements of the documentation, please feel free to point me to the spot or to open a PR.

In general, skorch shouldn't care much about what the `dataset.__getitem__` returns, _as long as the result from `get_iterator` consists of 2 elements_. Note that `get_iterator` does not even have to be a `DataLoader`, but usually you shouldn't mess with that.

Looking at the code now, I can envision that we might even be able to get rid of the requirement to have exactly 2 elements (but we would need to find some new solutions, e.g. how to determine the batch size). I think this is an interesting idea that could make skorch more universally usable.

> there can be multiple captions per image. Mandating a dictionary input whose values all have the same length means copying a lot of extra image data

Could you get into more details about how your data looks like? Maybe I can find a good suggestion how I would deal with the data.
